rust/lsp-lib/src/conversion_utils.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Utility functions meant for converting types from LSP to Core format
+//! and vice-versa
+
+use lsp_types::*;
+use xi_plugin_lib::{
+    Cache, Error as PluginLibError, Hover as CoreHover,
+    LanguageResponseError, Range as CoreRange,
+    PositionEnum, View
+};
+
+pub(crate) fn marked_string_to_string(marked_string: &MarkedString) -> String {
+    match *marked_string {
+        MarkedString::String(ref text) => text.to_owned(),
+        MarkedString::LanguageString(ref d) => format!("```{}\n{}\n```", d.language, d.value),
+    }
+}
+
+pub(crate) fn markdown_from_hover_contents(
+    hover_contents: HoverContents,
+) -> Result<String, LanguageResponseError> {
+    let res = match hover_contents {
+        HoverContents::Scalar(content) => marked_string_to_string(&content),
+        HoverContents::Array(content) => {
+            let res: Vec<String> = content.iter().map(|c| marked_string_to_string(c)).collect();
+            res.join("\n")
+        }
+        HoverContents::Markup(content) => content.value,
+    };
+    if res.is_empty() {
+        Err(LanguageResponseError::FallbackResponse)
+    } else {
+        Ok(res)
+    }
+}
+
+/// Counts the number of utf-16 code units in the given string.
```

Comment:
can we delete all this now?

Reply:
Actually, I currently handled the conversion for Hover only. This is still being used for the update methods. I plan to address them in a separate PR to keep it clean.